lib/cerner/oauth1a/access_token_agent.rb

```diff
@@ -143,6 +146,26 @@ def retrieve(principal: nil, ignore_cache: false)
         access_token
       end
 
+      # Public: Constructs an AccessToken using the realm of this agent and the value of an HTTP
+      # Authorization Header based on the OAuth HTTP Authorization Scheme (https://oauth.net/core/1.0a/#auth_header).
+      #
+      # value - A String containing the HTTP Authorization Header value.
+      #
+      # Returns an AccessToken.
+      #
+      # Raises a Cerner::OAuth1a::OAuthError with a populated oauth_problem and realm of this agent
+      # if any of the parameters in the value are invalid.
+      def from_authorization_header(value)
+        params = Protocol.parse_authorization_header(value)
+
+        unless params[:realm].nil? || params[:realm]&.eql?(@realm)
+          raise OAuthError.new('', nil, 'realm_rejected', nil, @realm)
+        end
+
+        # TODO: Is there a better way to get the realm into the token if it's not in the request header?
+        AccessToken.from_authorization_header(value, @realm)
```

Comment:
I'm not sure if there is a better way to give the agent's realm context to an access token being parsed from a header. For now, this just counts on an additional optional realm parameter in the `AccessToken.from_authorization_header` method.

Done this way, all realm validation happens here in the agent and the token treats any realm it has as being accurate. This means that anyone creating a token via `AccessToken.from_authorization_header` could end up with any arbitrary realm sent in the request header set into the www-authenticate header. To avoid this providers should always parse tokens via the agent.